cmd/kflex/adopt/adopt.go

```diff
@@ -0,0 +1,224 @@
+
+	// Construct a new kubeConfig object
+	kubeConfig := api.NewConfig()
+
+	kubeConfig.Clusters[context.Cluster] = cluster
+
+	if !skipURLOverride {
+		// Determine the server endpoint
+		endpoint := adoptedURLOverride
+		if endpoint == "" {
+			endpoint = cluster.Server
+			if !isValidServerURL(endpoint) {
+				return fmt.Errorf("invalid server endpoint %s. Please provide a valid value with the `url-override` option", endpoint)
+			}
+		}
+		kubeConfig.Clusters[context.Cluster].Server = endpoint
+	}
+
+	if authInfo, exists := config.AuthInfos[context.AuthInfo]; exists {
+		kubeConfig.AuthInfos[contextName] = authInfo
+	} else {
+		return fmt.Errorf("authInfo %s not found for context %s", context.AuthInfo, contextName)
+	}
+
+	kubeConfig.Contexts[contextName] = &api.Context{
+		Cluster:  context.Cluster,
+		AuthInfo: contextName,
+	}
+	kubeConfig.CurrentContext = contextName
+
+	newKubeConfig, err := clientcmd.Write(*kubeConfig)
+	if err != nil {
+		return fmt.Errorf("failed to serialize the new kubeconfig: %v", err)
+	}
+
+	createOrUpdateSecret(clientset, cpName, newKubeConfig)
+
+	return nil
+}
+
+func createOrUpdateSecret(clientset *kubernetes.Clientset, cpName string, kubeconfig []byte) error {
+
+	// Define the kubeconfig secret
+	kubeConfigSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateBoostrapSecretName(cpName),
+			Namespace: util.SystemNamespace,
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: map[string][]byte{util.KubeconfigSecretKeyInCluster: kubeconfig},
+	}
+
+	// Try to create the secret
+	if _, err := clientset.CoreV1().Secrets(util.SystemNamespace).Create(context.TODO(), kubeConfigSecret, metav1.CreateOptions{}); err != nil {
+		// Check if the error is because the secret already exists
+		if apierrors.IsAlreadyExists(err) {
+			// Retrieve the existing secret
+			existingSecret, getErr := clientset.CoreV1().Secrets(util.SystemNamespace).Get(context.TODO(), util.AdminConfSecret, metav1.GetOptions{})
+			if getErr != nil {
+				return fmt.Errorf("failed to fetch existing secret %s in namespace %s: %v", util.AdminConfSecret, util.SystemNamespace, getErr)
+			}
+
+			// Update the data of the existing secret
+			existingSecret.Data = kubeConfigSecret.Data
+
+			// Update the secret with new data
+			if _, updateErr := clientset.CoreV1().Secrets(util.SystemNamespace).Update(context.TODO(), existingSecret, metav1.UpdateOptions{}); updateErr != nil {
+				return fmt.Errorf("failed to update existing secret %s in namespace %s: %v", util.AdminConfSecret, util.SystemNamespace, updateErr)
+			}
+		} else {
+			return fmt.Errorf("failed to create secret %s in namespace %s: %v", util.AdminConfSecret, util.SystemNamespace, err)
+		}
+	}
+
+	return nil
+}
+
+// check if the current server URL in the adopted cluster kubeconfig is using
+// a local address, which would not work in a container
+func isValidServerURL(serverURL string) bool {
+	localAddresses := []string{"127.0.0.1", "localhost", "::1"}
```

Comment:
`strings.Contains("2600:0062::1", "::1")` is true.
`strings.Contains("nonlocalhost", "localhost")` is true.
`strings.Contains("127.0.0.100", "127.0.0.1")` is true.